siff-dr/siff-header.go

```diff
@@ -35,28 +35,49 @@ func setSiffFields(packet *netfilter.NFPacket, flags layers.IPv4Flag, capabiliti
 	}
 
 	/* Modify the ip layer information */
-	// compute new IHL and length
-	var IHLchange uint8 = (*ipLayer).IHL
+	var IHLchange uint16 = uint16((*ipLayer).IHL)
 
+	// compute new IHL and length
 	if (flags & 0x03) == IS_SIFF {
-		(*ipLayer).IHL = 7
-	} else if (flags & 0x03) == (IS_SIFF | CAPABILITY_UPDATE) {
 		(*ipLayer).IHL = 9
+	} else if (flags & 0x03) == (IS_SIFF | CAPABILITY_UPDATE) {
+		(*ipLayer).IHL = 11
+	} else {	// evil
+		(*ipLayer).IHL = 7
 	}
 
-	/* change the total length by the change in IHL * 4 to convert from
-	   32-bit words to bytes */
-	IHLchange = (*ipLayer).IHL - IHLchange
+	IHLchange = uint16((*ipLayer).IHL) - IHLchange
 	if IHLchange != 0 {
-		(*ipLayer).Length = uint16((*ipLayer).IHL) * 4
+		(*ipLayer).Length = (*ipLayer).Length + IHLchange * 4
 	}
 
-	// set the flags, preserving the first flag bit in case it is used
-	(*ipLayer).IHL = flags
+	/* change the total length by the change in IHL * 4 to convert from
+	   32-bit words to bytes */
+	(*ipLayer).Length = uint16((*ipLayer).IHL) * 4
+
+	// set the flags option
+	var flagOption layers.IPv4Option
+	flagOption.OptionType = 86
+	flagOption.OptionLength = 8
+	var flag_array [6]byte = [6]byte{0, 0, 0, 0, 0, 0}
```

Comment:
Yeah, I think if we use the evil bit from gopacket, we wouldn't even need
options for legacy packets. In theory we could even reduce it to one flag
that way, but I think as long as we keep the flags to 2 bytes, it should be
acceptable. I know he isn't looking at our code, but we should at least
pretend we tried to address overhead so we don't have to completely lie
about it in the paper.

On Sun, Dec 13, 2015 at 3:18 PM Alec Thompson <notifications@github.com>
wrote:

> In siff-dr/siff-header.go
> <https://github.com/ThomasJClark/cs4516project/pull/21#discussion_r47449537>
> :
>
> >  	}
> >
> > -	// set the flags, preserving the first flag bit in case it is used
> > -	(*ipLayer).IHL = flags
> > +	/* change the total length by the change in IHL * 4 to convert from
> > +	   32-bit words to bytes */
> > +	(*ipLayer).Length = uint16((*ipLayer).IHL) * 4
> > +
> > +	// set the flags option
> > +	var flagOption layers.IPv4Option
> > +	flagOption.OptionType = 86
> > +	flagOption.OptionLength = 8
> > +	var flag_array [6]byte = [6]byte{0, 0, 0, 0, 0, 0}
>
> I guess we could just use one option for all of it, and there would be
> less overhead, but the way it's being handled now is that all packets
> should have the flags option, siff packets have flags and the capabilities
> option, and siff packets with updates have all three.
>
> And I wasn't aware the evil bit was defined in the go package structure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ThomasJClark/cs4516project/pull/21/files#r47449537>.
>